src/main/java/com/github/cameltooling/dap/internal/CamelDebugAdapterServer.java

```diff
@@ -301,4 +334,49 @@ public CompletableFuture<Void> configurationDone(ConfigurationDoneArguments args
 			}
 		);
 	}
+
+	/**
+	 * Executes asynchronously the given task ensuring that the context class loader is properly set to ensure that
+	 * the classes from third party libraries are found.
+	 *
+	 * @param runnable the task to execute
+	 * @return the new CompletableFuture
+	 */
+	private static CompletableFuture<Void> runAsync(Runnable runnable) {
+		final ClassLoader callerCCL = Thread.currentThread().getContextClassLoader();
+		return CompletableFuture.runAsync(
+			() -> {
+				final ClassLoader currentCCL = Thread.currentThread().getContextClassLoader();
+				try {
+					Thread.currentThread().setContextClassLoader(callerCCL);
+					runnable.run();
+				} finally {
+					Thread.currentThread().setContextClassLoader(currentCCL);
+				}
+			}
+		);
+	}
+
+	/**
+	 * Calls asynchronously the given supplier ensuring that the context class loader is properly set to ensure that
+	 * the classes from third party libraries are found.
+	 *
+	 * @param supplier the supplier to call
+	 * @return the new CompletableFuture
+	 * @param <U> the type of the result
+	 */
+	private static <U> CompletableFuture<U> supplyAsync(Supplier<U> supplier) {
+		final ClassLoader callerCCL = Thread.currentThread().getContextClassLoader();
+		return CompletableFuture.supplyAsync(
+			() -> {
+				final ClassLoader currentCCL = Thread.currentThread().getContextClassLoader();
+				try {
+					Thread.currentThread().setContextClassLoader(callerCCL);
+					return supplier.get();
+				} finally {
+					Thread.currentThread().setContextClassLoader(currentCCL);
+				}
+			}
+		);
+	}
```

Comment:
Can you precise the use case of the Third-party libraries?

I tried to remove the classloader modifications and all tests are passing fine.
By experience, touching the classloaders is very often a source of problem so would be sure to understand why it is needed here.

Reply:
To reproduce, you need to build and redeploy the extension. Then try to debug a project, you should get an error of the next type in your log file:

```
2022-07-11 15:33:40.691 ERROR [Camel DAP - Check Suspended node] [com.github.cameltooling.dap.internal.types.UnmarshallerEventMessage.getUnmarshalledEventMessage(UnmarshallerEventMessage.java:38)] - Cannot parse message from debugger
javax.xml.bind.JAXBException: Implementation of JAXB-API has not been found on module path or classpath.
	at javax.xml.bind.ContextFinder.newInstance(ContextFinder.java:232) ~[jakarta.xml.bind-api-2.3.3.jar!/:2.3.3]
	at javax.xml.bind.ContextFinder.find(ContextFinder.java:375) ~[jakarta.xml.bind-api-2.3.3.jar!/:2.3.3]
	at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:691) ~[jakarta.xml.bind-api-2.3.3.jar!/:2.3.3]
	at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:632) ~[jakarta.xml.bind-api-2.3.3.jar!/:2.3.3]
	at com.github.cameltooling.dap.internal.types.UnmarshallerEventMessage.getUnmarshalledEventMessage(UnmarshallerEventMessage.java:34) ~[classes!/:?]
	at com.github.cameltooling.dap.internal.BacklogDebuggerConnectionManager.handleSuspendedBreakpoint(BacklogDebuggerConnectionManager.java:197) ~[classes!/:?]
	at com.github.cameltooling.dap.internal.BacklogDebuggerConnectionManager.checkSuspendedBreakpoints(BacklogDebuggerConnectionManager.java:178) ~[classes!/:?]
	at java.lang.Thread.run(Thread.java:829) ~[?:?]
Caused by: java.lang.ClassNotFoundException: com.sun.xml.bind.v2.ContextFactory
	at jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581) ~[?:?]
	at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:522) ~[?:?]
	at javax.xml.bind.ServiceLoaderUtil.nullSafeLoadClass(ServiceLoaderUtil.java:92) ~[jakarta.xml.bind-api-2.3.3.jar!/:2.3.3]
	at javax.xml.bind.ServiceLoaderUtil.safeLoadClass(ServiceLoaderUtil.java:125) ~[jakarta.xml.bind-api-2.3.3.jar!/:2.3.3]
	at javax.xml.bind.ContextFinder.newInstance(ContextFinder.java:230) ~[jakarta.xml.bind-api-2.3.3.jar!/:2.3.3]
	... 7 more
```

This is due to the fact that, the `spring-boot-loader` isolates the classes of the extension and the third party libraries into a `LaunchedURLClassLoader` while the threads executing the code (from the common pool) are initialized with  AppClassLoader such that they cannot see the classes of the extension thus we end up with CNFE.

There are 2 ways to fix it, either we set explicitly the CCL like in this PR or we use a specific executor (thread pool) to ensure that the threads will be initialized with the same CL as the extension code.

Reply:
@apupier Is-it clear enough or should I clarify more?

Reply:
sorry, not sure when I will have time to have a look. I will really try today.

Reply:
I feel like in Debug Adapter Server, we will always want to have all dependencies in the classloader. So using a specific executor sounds to be better matching our case. It will avoid to forget to use the provided functions setting the ClassLoader when developing next features.
This can be done in another Pull request. If you agree, I will raise a specific issue to do it later.

Reply:
Please note that in both cases, you will need to do something specific for each asynchronous task, either you need to set the CCL (with this approach) or force the executor (with other approach) so you likely need utility methods in both cases.

Reply:
:-o
This really sounds error-prone. Also I'm failing to see why by default there is this behavior on the ClassLoaders for the threads.

Reply:
If you don't specify any executor to the Async methods of `CompletableFuture`, by default it uses the common pool  from `ForkJoinPool.commonPool()`. If you don't provide any thread factory to `ForkJoinPool` by default it uses a thread factory that initializes the threads of the pool with the system class loader as CCL.

As you can see [in the Javadoc](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ForkJoinPool.html), I quote:

> If no thread factory is supplied via a system property, then the common pool uses a factory that uses the system class loader as the [thread context class loader](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Thread.html#getContextClassLoader()).